tests/unit/s2n_self_talk_inet_socket_test.c

```diff
@@ -0,0 +1,187 @@
+            S2N_DEFAULT_TEST_CERT_CHAIN, S2N_DEFAULT_TEST_PRIVATE_KEY));
+
+    /* Setup config */
+    EXPECT_SUCCESS(s2n_connection_set_blinding(server_conn, S2N_SELF_SERVICE_BLINDING));
+    EXPECT_EQUAL(s2n_connection_get_delay(server_conn), 0);
+    EXPECT_SUCCESS(s2n_connection_set_fd(server_conn, fd));
+    EXPECT_SUCCESS(s2n_config_set_cipher_preferences(config, "default"));
+    EXPECT_SUCCESS(s2n_config_set_unsafe_for_testing(config));
+    EXPECT_SUCCESS(s2n_config_add_cert_chain_and_key_to_store(config, chain_and_key));
+    EXPECT_SUCCESS(s2n_connection_set_config(server_conn, config));
+
+    /* Complete the handshake */
+    s2n_blocked_status blocked = S2N_NOT_BLOCKED;
+    EXPECT_SUCCESS(s2n_negotiate(server_conn, &blocked));
+    EXPECT_EQUAL(server_conn->actual_protocol_version, S2N_TLS12);
+
+    RESULT_GUARD(socket_cb->server_post_handshake_cb(server_conn));
+
+    char sync = 0;
+    char send_buffer[1] = { 0 };
+    {
+        send_buffer[0] = CHAR_A;
+        EXPECT_SUCCESS(s2n_send(server_conn, send_buffer, 1, &blocked));
+        EXPECT_SUCCESS(write(write_pipe, &sync, 1));
+
+        send_buffer[0] = CHAR_B;
+        EXPECT_SUCCESS(s2n_send(server_conn, send_buffer, 1, &blocked));
+        EXPECT_SUCCESS(write(write_pipe, &sync, 1));
+    }
+
+    return S2N_RESULT_OK;
+}
+
+static S2N_RESULT launch_test(const struct self_talk_inet_socket_callbacks *socket_cb)
+{
+    /* configure real socket */
+    int listener = socket(AF_INET, SOCK_STREAM, 0);
+    RESULT_GUARD_POSIX(listener);
+    struct sockaddr_in saddr = { 0 };
+    saddr.sin_family = AF_INET;
+    saddr.sin_addr.s_addr = htonl(S2N_TEST_INADDR_LOOPBACK);
+    saddr.sin_port = 0;
+
+    /* listen on socket address */
+    socklen_t addrlen = sizeof(saddr);
+    RESULT_GUARD_POSIX(bind(listener, (struct sockaddr *) &saddr, addrlen));
+    RESULT_GUARD_POSIX(getsockname(listener, (struct sockaddr *) &saddr, &addrlen));
+
+    /* Used for synchronizing read and writes between client and server.
+     *
+     * The purpose of this mechanism is to prevent the client from reading
+     * before the server has sent it. The client blocks on reading a single
+     * char while the server writes a char after it has sent data and wishes
+     * to unblock the client. */
+    int sync_pipe[2] = { 0 };
+    RESULT_GUARD_POSIX(pipe(sync_pipe));
+
+    pid_t child = fork();
+    RESULT_ENSURE(child >= 0, S2N_ERR_SAFETY);
+    int status = 0;
+    int fd = 0;
+    if (child) {
+        /* server */
+        RESULT_GUARD_POSIX(listen(listener, 1));
+        fd = accept(listener, NULL, NULL);
+        RESULT_GUARD_POSIX(fd);
+
+        RESULT_GUARD_POSIX(close(sync_pipe[0]));
+        RESULT_GUARD(start_server(fd, sync_pipe[1], socket_cb));
+
+        RESULT_ENSURE_EQ(waitpid(-1, &status, 0), child);
+        RESULT_ENSURE_EQ(status, 0);
+    } else {
+        /* client */
+        fd = socket(AF_INET, SOCK_STREAM, 0);
```

Comment:
Won't you need some way to configure the sockets that `launch_test` uses for kTLS?

Reply:
That will happen via the callbacks: in future PRs https://github.com/toidiu/s2n-tls/blob/b7aa689c11898df970f670c851e4b8e0a262991e/tests/testlib/s2n_self_talk_test_utils.c